### ISSUE TITLE
apply_patch function failed for array in nested structure

### ISSUE THREAD

**Issue by kevinxin90** [stefankoegl/python-json-patch#41]
This issue is similar to issue #40 , but it happens when using the apply_patch function.

A 'JsonPatchConflict: can't insert outside of list' error would raise every time we try to use apply_patch function between two docs having array inside a nested structure.

Take the following example:

```
doc5 = {'school':{'names':['Kevin','Carl']}}

doc6 = {'school':{'names':['Carl','Kate','Kevin','Jake']}}

patch = jsonpatch.JsonPatch.from_diff(doc5, doc6)

jsonpatch.apply_patch(doc5,patch)
```

When it run it produce the following error:

```
----> 1 jsonpatch.apply_patch(doc5,patch)

/home/kevinxin/ENV/src/jsonpatch/jsonpatch.py in apply_patch(doc, patch, in_place)
    149     else:
    150         patch = JsonPatch(patch)
--> 151     return patch.apply(doc, in_place)
    152
    153

/home/kevinxin/ENV/src/jsonpatch/jsonpatch.py in apply(self, obj, in_place)
    347
    348         for operation in self._ops:
--> 349             obj = operation.apply(obj)
    350
    351         return obj

/home/kevinxin/ENV/src/jsonpatch/jsonpatch.py in apply(self, obj)
    498             'path': self.location,
    499             'value': value
--> 500         }).apply(obj)
    501
    502         return obj

/home/kevinxin/ENV/src/jsonpatch/jsonpatch.py in apply(self, obj)
    422
    423             elif part > len(subobj) or part < 0:
--> 424                 raise JsonPatchConflict("can't insert outside of list")
    425
    426             else:

JsonPatchConflict: can't insert outside of list
```

However, it the array is not in nested structure, the problem is gone.

Take the following example：

```
doc3 = {'names':['kevin','Carl']}

doc4 = {'names':['Carl','Kate','Kevin','Jake']}

patch = jsonpatch.JsonPatch.from_diff(doc3, doc4)

jsonpatch.apply_patch(doc3,patch)
Out: {'names': ['Carl', 'Kate', 'Kevin', 'Jake']}
```

**Comment by nxsofsys** [comment]
This issue is related to diff construction. You can try https://github.com/nxsofsys/jsondiff instead builtin diff constructor.

**Comment by kevinxin90** [comment]
Hi,

Super helpful! Fix the problem! Thanks a lot!!

Kevin

**Comment by kevinxin90** [comment]
version 1.12 still didn't fix this issue, get same error as before. For the moment, still using jsondiff.py as a workaround.

**Comment by woosley** [comment]
@kevinxin90 I met the same issue, how can I use jsondiff here?

**Comment by selurvedu** [comment]
Me too. I've got 5 testcases [here](https://github.com/selurvedu/python-json-patch/compare/list_tests@v1.13) which fail with vanilla python-json-patch but pass with jsondiff.

**Comment by stefankoegl** [comment]
@selurvedu I have created a pull request for your commits in #68.

**Comment by stefankoegl** [comment]
I have just released jsonpatch 1.20 which fixes the issue you describe. In addition I have added a test for this case in ae895f7ed9aa96e96d9d3ba402d3e2d0a11b0c8e.

Thanks for reporting!
